crates/ide-assists/src/handlers/change_visibility.rs

```diff
@@ -46,16 +47,36 @@ fn add_vis(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
 
     let (offset, target) = if let Some(keyword) = item_keyword {
         let parent = keyword.parent()?;
-        let def_kws =
-            vec![CONST, STATIC, TYPE_ALIAS, FN, MODULE, STRUCT, ENUM, TRAIT, USE, MACRO_DEF];
-        // Parent is not a definition, can't add visibility
-        if !def_kws.iter().any(|&def_kw| def_kw == parent.kind()) {
+
+        let can_add = |node: &SyntaxNode| -> bool {
+            const LEGAL: &[syntax::SyntaxKind] =
+                &[CONST, STATIC, TYPE_ALIAS, FN, MODULE, STRUCT, ENUM, TRAIT, USE, MACRO_DEF];
+
+            LEGAL.contains(&node.kind()) && {
+                let is_legal = |k| matches!(k, SOURCE_FILE | MODULE);
+                let Some(p) = node.parent() else {
+                    return false;
+                };
+
+                if p.kind() == ASSOC_ITEM_LIST {
+                    p.parent()
+                        .filter(|imp| imp.kind() == IMPL)
+                        .and_then(|imp| imp.parent())
+                        .is_some_and(|it| is_legal(it.kind()))
```

Comment:
Can you check if trait impls (i.e. `impl Trait for Type`) don't bypass this check? They are also parsed as `IMPL` and we don't want to provide this assist for them I suppose.

Reply:
Ah, can you actually exempt inherent (non-trait) impls from this check and make them "legal" even when they're not at top-level? Impls have non-local effect and it's reasonable to provide the assist for cases like this:

```rust
const _: () = {
    impl S {
        fn/*$0*/ foo() {} 
    }
};
```